File: SRM05B/SRM05B.py
import re
t = int(input())
ans = []
for i in range(0,t):
    ans.append("invalid")
    ip = input()
    if "." in ip:
        parts = ip.split(".")
        if len(parts) == 4:
            for j in range(0,len(parts)):
                if len(parts[j]) == 0:
                    ans[i] = "invalid"
                    break
                else:
                    if parts[j][0] == '0':
                        if len(parts[j]) == 1:
                            if j < 3:
                                continue
                            elif j == 3:
                                ans[i] = "ipv4"
                        else:
                            ans[i] = "invalid"
                            break
                    if parts[j].isnumeric() and len(parts[j]) <= 3:
                        temp = int(parts[j])
                        if temp >=0 and temp<=255:
                            if j < 3:
                                continue
                            elif j == 3:
                                ans[i] = "ipv4"
                        else:
                            ans[i] = "invalid"
                            break
                    else:
                        ans[i] = "invalid"
                        break
        else:
            ans[i] = "invalid"
    elif ":" in ip:
        parts = ip.split(":")
        if len(parts) == 8:
            for j in range(0,len(parts)):
                if parts[j][0] == '0':
                    if len(parts[j]) == 1:
                        if j < 7:
                            continue
                        elif j == 7:
                            ans[i] = "ipv6"
                    else:
                        ans[i] = "invalid"
                        break
                if bool(re.match('^[a-f0-9]+$', parts[j])) == True and len(parts[j]) <= 4:
                    if j < 7:
                        continue
                    elif j == 7:
                        ans[i] = "ipv6"
                else:
                    ans[i] = "invalid"
                    break
        else:
            ans[i] = "invalid"
    else:
        ans[i] = "invalid"
for i in ans:
    print(i)

File: SRM05B/Srm05B.java
import java.util.*;
import java.lang.*;
import java.io.*;
class Srm05B
{
	public static void main (String[] args) throws java.lang.Exception
	{
		Scanner sc = new Scanner(System.in);
		int tCases = sc.nextInt();
		sc.nextLine();
		String ans[] = new String[tCases];
		for(int i = 0 ; i < tCases ; i++) {
		    String ip = sc.nextLine();
		    if(ip.indexOf('.') >0) {
		        String parts[] = ip.split("\\.");
		        if(parts.length != 4)
		            ans[i] = "invalid";
		        else {
		            for(int j = 0 ; j < 4 ; j++) {
		                if(parts[j].length() == 0 ) {
		                    ans[i] = "invalid";
		                    break;
		                }
		                else
    		            {
    		                if( parts[j].charAt(0) == '0') {
    		                    if(parts[j].length() == 1) {
    		                        if(j<3)
                                        continue;
                                    else if(j == 3)
                                        ans[i] = "ipv4";
    		                    }
    		                    else {
    		                        ans[i] = "invalid";
    		                        break;
    		                    }
    		                }
    		                if (parts[j].matches("[0-9]+") && parts[j].length() <= 3) {
                                int temp = Integer.parseInt(parts[j]);
                                if(temp >=0 && temp <= 255) {
                                    if(j<3)
                                        continue;
                                    else if(j == 3)
                                        ans[i] = "ipv4";
                                }
                                else {
                                    ans[i] = "invalid";
                                    break;
                                }
                            }
                            else {
                                ans[i] = "invalid";
                                break;
                            }
    		            }  
		            }
		        }
		    }
		    else if(ip.indexOf(':') > 0) {
		        String parts[] = ip.split("\\:");
		        if(parts.length != 8)
		            ans[i] = "invalid";
		        else {
		            for(int j = 0 ; j < 8 ; j++) {
		                if(parts[j].charAt(0) == '0') {
		                    if(parts[j].length() == 1) {
		                        if(j<7)
    		                        continue;
    		                    else if(j==7)
    		                        ans[i] = "ipv6";
		                    }
		                    else {
		                        ans[i] = "invalid";
		                        break;
		                    }
		                }
		                if(parts[j].matches("[0-9a-f]+") && parts[j].length() <= 4) {
		                    if(j<7)
		                        continue;
		                    else if(j==7)
		                        ans[i] = "ipv6";
		                }
		                else {
		                    ans[i] = "invalid";
		                    break;
		                }
		            }
		        }
		    }
		    else
		        ans[i] = "invalid";
		}
		for(String i : ans)
			System.out.println(i);
	}
}